contracts/test/Staking.t.sol

```diff
@@ -0,0 +1,1399 @@
+
+        vm.prank(staker2);
+        token.approve(address(staking), type(uint256).max);
+    }
+
+    // ============================================================
+    // HELPER FUNCTIONS
+    // ============================================================
+
+    /// @dev Internal function to update validator registration status
+    function _updateValidator(address _validator, bool _isRegistration) internal {
+        vm.startPrank(owner);
+        address[] memory validators = new address[](1);
+        validators[0] = _validator;
+        bool[] memory isRegistration = new bool[](1);
+        isRegistration[0] = _isRegistration;
+        staking.proposeValidators(validators, isRegistration);
+        vm.warp(block.timestamp + CONFIG_TIME_DELAY);
+        staking.executeValidatorChanges(validators, isRegistration, block.timestamp);
+        vm.stopPrank();
+    }
+
+    /// @dev Registers a new validator through the timelock process
+    function _registerValidator(address _validator) internal {
+        _updateValidator(_validator, true);
+    }
+
+    /// @dev Deregisters a validator through the timelock process
+    function _deregisterValidator(address _validator) internal {
+        _updateValidator(_validator, false);
+    }
+
+    /// @dev Stakes tokens as a specific staker to a validator
+    function _stakeAs(address _staker, address _validator, uint256 _amount) internal {
+        vm.prank(_staker);
+        staking.stake(_validator, _amount);
+    }
+
+    /// @dev Initiates a withdrawal and returns the withdrawal ID
+    function _initiateWithdrawalAs(address _staker, address _validator, uint256 _amount)
+        internal
+        returns (uint64 expectedId)
+    {
+        expectedId = staking.nextWithdrawalId();
+        vm.prank(_staker);
+        staking.initiateWithdrawal(_validator, _amount);
+    }
+
+    /// @dev Asserts complete state consistency for a staker-validator pair
+    function _assertStakeState(
+        address _staker,
+        address _validator,
+        uint256 expectedStake,
+        uint256 expectedTotalValidatorStake,
+        uint256 expectedTotalStakerStake
+    ) internal view {
+        assertEq(staking.stakes(_staker, _validator), expectedStake);
+        assertEq(staking.totalValidatorStakes(_validator), expectedTotalValidatorStake);
+        assertEq(staking.totalStakerStakes(_staker), expectedTotalStakerStake);
+    }
+
+    // ============================================================
+    // CONFIGURATION & ACCESS CONTROL
+    // ============================================================
+
+    function test_Constructor() public view {
+        assertEq(address(staking.SAFE_TOKEN()), address(token));
+        assertEq(staking.CONFIG_TIME_DELAY(), CONFIG_TIME_DELAY);
+        assertEq(staking.withdrawDelay(), INITIAL_WITHDRAW_DELAY);
+        assertEq(staking.owner(), owner);
+        assertTrue(staking.isValidator(validator));
```

Comment:
strictly speaking this is not part of the constructor. The constructor should not set any validator, right?

Reply:
d1f019b